src/Plugin/Commerce/PaymentGateway/MidtransInstallment.php

```diff
@@ -109,61 +111,126 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
-    // $logger = \Drupal::logger('commerce_midtrans');    
-    drupal_set_message('Thank you for placing your order'); 
+    $this->midtransConfig();
+    $payment = $this->loadPaymentByOrderId($order->id());
+    $response = \Midtrans\Transaction::status($payment->getRemoteId());
+
+    $installment = isset($response->installment_term) ? ' - Installment: '.$response->installment_term. ' Months' : '';
+    $payment_type = ucwords($response->card_type).' Card - Mask Card: '.$response->masked_card.$installment;
+
+    $message = '<p><strong>Here is the detail payment from Midtrans</strong><br />';
+    $message .= 'Order ID: '.$response->order_id.'<br>';
+    $message .= 'Transaction ID: '.$response->transaction_id.'<br>';
+    $message .= 'Transaction Status: '.ucwords($response->transaction_status).'<br>';
+    $message .= 'Payment Type: '.$payment_type.'<br>';
+
+    $this->messenger()->addMessage($this->t($message));
+  }
+
+  public function onCancel(OrderInterface $order, Request $request) {
+    // clear snap token, if snap token was expired, can use new snap token
+    $order->setData('snap_token', NULL);
+    $order->save();
+
+    $this->messenger()->addMessage($this->t('You have canceled checkout at @gateway but may resume the checkout process here when you are ready.', [
+      '@gateway' => $this->getDisplayLabel(),
+    ]));
   }
 
   /**
    * {@inheritdoc}
    */
   public function onNotify(Request $request) {
-    \Midtrans\Config::$serverKey =  $this->getConfiguration()['server_key'];
-    \Midtrans\Config::$isProduction = ($this->getMode() == 'production') ? TRUE : FALSE;
-    $response = new \Midtrans\Notification();
-    /** @var \Drupal\commerce_payment\PaymentStorage $payment_storage */
-    $payment_storage = $this->entityTypeManager->getStorage('commerce_payment');
-    /** @var \Drupal\commerce_payment\Entity\Payment $payment */
-    //$payment = $payment_storage->loadByRemoteId($response->order_id);
-    /** @var \Drupal\commerce_order\Entity\Order $order */
-    //$order = $payment->getOrder();
-    
-    //error_log('Response from Midtrans : '. print_r($response, TRUE)); //debugan  
-    $payment = $this->loadPaymentByOrderId1($response->order_id);
+    $this->midtransConfig();
+    $notification = new \Midtrans\Notification();
+    $response = $notification->getResponse();
+    $payment = $this->loadPaymentByOrderId($response->order_id);
+    $order = $payment->getOrder();
+
+    // change payment remote id with transaction id
+    $payment->setRemoteId($response->transaction_id);
+    $payment->save();
+
+    $message = 'orderID '.$response->order_id.' - '.$response->payment_type.' - '.$response->transaction_status;
+    \Drupal::logger('commerce_midtrans')->info($message);
+
+    $installment = isset($response->installment_term) ? ' - Installment: '.$response->installment_term. ' Months' : '';
+    $payment_type = ucwords($response->card_type).' Card - Mask Card: '.$response->masked_card.$installment;
```

Comment:
where is this `$payment_type` var being used? looks like unused on this function.